Add ChoiceParameter to restrict parameter options to those specified.

brcopeland opened the pull request:
<!--- This template is optional. Please use it as a starting point to help guide PRs -->

<!--- Provide a general summary of your changes in the Title above -->

## Description
As in description.

## Motivation and Context
This allows the user to define the permissible values for a parameter and the value type, and reject any parameter that does not meet these criteria.

## Have you tested this? If so, how?
I've been using the class for several weeks without issue.

erikbern commented:
lgtm although would be cool if this works with `--help`

brcopeland commented:
@erikbern, can you give me an example of what you mean?  I'd be happy to make the change(s).  When I create a Task using a ChoiceParameter, the parameter and its description show up when I run `luigi --module <module> <Task> --help`.

erikbern commented:
would be cool if the `--help` also shows what options are available. but no biggie :)

erikbern commented on luigi/parameter.py:
```diff
@@ -959,3 +959,21 @@ def serialize(self, x):
+        if var in self.choices:
+            return var
+        else:
+            raise ValueError("{s} is not a valid choice from {choices}".format(
```
this line is broken – can you fix

brcopeland replied:
Yes, fixed.  Don't know how I didn't copy that last line.

brcopeland commented:
Ah, yeah.  I'm trying to figure out a way to do that - updating self.description in __init__ does not appear to be sufficient.

brcopeland commented:
Ok, got it working.  My system has a luigi install I was importing instead of the one I was editing.

erikbern commented:
awesome! flake8 is complaining: https://travis-ci.org/spotify/luigi/jobs/149050059

Tarrasch commented:
Can you also add test cases please?

Tarrasch commented on luigi/parameter.py:
```diff
@@ -959,3 +959,28 @@ def serialize(self, x):
+    """
+    def __init__(self, choices, var_type=str, *args, **kwargs):
+        super(ChoiceParameter, self).__init__(*args, **kwargs)
+        self.choices = set(var_type(choice) for choice in choices)
```
Can you make this into an assertion? I imagine people creating new `Parameter`s know enough luigi+python that they would rather be explicit in types over than luigi magically converting stuff behind their back.

Tarrasch commented on luigi/parameter.py:
```diff
@@ -959,3 +959,28 @@ def serialize(self, x):
+    """Parameter that restricts options to the specified set
+    All choices must be of the same type that so that parse can
+    correctly infer type
+    Variable type is specified with var_type=...
```
Please format all docs and make sure it renders nice. See CONTRIBUTING.md and other examples in this file.

Tarrasch commented:
This class is somewhat similar to `EnumParameter`. Would you mind interlinking those two in the docs? Maybe add something like "If you want a more typed and structured version of a ChoiceParameter, you should consider using an EnumParameter instead." could guide our users to pick the right choice?

brcopeland commented:
You could be entirely right here.  I am a pretty new user of luigi and may have overlooked this class before coming up with my own version.  I will clean it up if requested, but maybe at this point it just makes more sense to not include this.

Tarrasch commented:
I still think this makes sense. Not everyone likes to create a new Enum type for every input-restriction and add all the conversion overhead if they want to eventually go back to a string type. I would be happy to include this, as long as the docs comes with recommended good practices. Like we've previously  done [here](https://github.com/ehdr/luigi/blob/f771622fe2b8cca39b1165c13a47f444f7c46b03/luigi/contrib/external_program.py#L24-L29) for example. :)

brcopeland commented:
I changed the docstring to I believe the proper format and added a test suite to the test/ directory.  Please let me know if any further changes would be welcome.

Tarrasch commented on luigi/parameter.py:
```diff
@@ -995,3 +995,49 @@ def parse(self, s):
+    def __init__(self, var_type=str, *args, **kwargs):
+        if "choices" not in kwargs:
+            raise ParameterException("A choices iterable must be specified")
+        self._choices = set(var_type(choice) for choice in kwargs.pop("choices"))
```
As I pointed out [earlier](https://github.com/spotify/luigi/pull/1800#discussion_r73090575), I probably would prefer if we could somehow would assert instead of convert. But I'm not sure really what's best :)

brcopeland replied:
Oh, ok sure I think that's perfectly reasonable.

Tarrasch commented:
Apart from the small comment about assert/convert. This looks *very good*. I must say I'm very impressed how you managed to fix all the shortcomings in a single review turnaround! :)

brcopeland commented:
Made a couple changes to assert the choices are the proper type as you suggest.  And sure, glad it's helpful.

brcopeland commented:
I rebased, updated docstrings, and added unit tests for the class.

Tarrasch commented:
Awesomeness! :)

Tarrasch commented:
A post-merge note. Some users might want to use the ["Functional API"](https://docs.python.org/3/library/enum.html#functional-api) of enums. Instead of 

    ChoiceParameter(choices=['a', 'b', 'c'])

Some users might prefer

    EnumParameter(enum=Enum('Character', 'a b c'))
